### PR TITLE
(BIDS-1458) ENS backend

### PR DESCRIPTION
**Pull request by MauserBitfly**
This PR will be replaced by https://github.com/gobitfly/eth2-beaconchain-explorer/pull/1987

I leave this one open (until the other one is merged), in case we decide not to go with the event approach.

### CODE REVIEW

**Review by recy21 — APPROVED** [review]

File: handlers/ens.go [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+func GetEnsDomain(search string) (*types.EnsDomainResponse, error) {
+	data := &types.EnsDomainResponse{}
+	var returnError error
+
+	if utils.IsValidEnsDomain(search) {
+		data.Domain = search
+		cacheKey := fmt.Sprintf("%d:ens:address:%v", utils.Config.Chain.Config.DepositChainID, search)
+
+		if address, err := cache.TieredCache.GetStringWithLocalTimeout(cacheKey, time.Hour); err != nil || len(address) == 0 {
```
Comment (D13ce) [inline handlers/ens.go:44]: As discussed with @peterbitfly, caching introduces inaccuracy as ENS names can change every block.
Removing caching with the implementation here would potentially mean a lot of traffic for our node.

I am afraid we cannot implement this whole feature the way you did. @peterbitfly suggested to index events on the according ENS contract in bigtable to always be up2date and to reduce the load on our nodes.

**Review by D13ce — CHANGES_REQUESTED** [review]
I am afraid we will not be able to follow this approach. I will link my comment explaining why as soon as I have submitted this review (i.e. in the next few seconds)

Furthermore, I have observed the following behavior after starting a local instance of this:

<http://localhost:8081/api/v1/ens/lookup/allice.eth>
```
{"status":"OK","data":{"address":"0x20a5bD183b420eBBD1BdA4b8818801b9eFa4CA37","domain":"allice.eth"}}
```

<http://localhost:8081/api/v1/ens/lookup/0x20a5bD183b420eBBD1BdA4b8818801b9eFa4CA37>
```
{"status":"OK","data":{"address":"0x20a5bD183b420eBBD1BdA4b8818801b9eFa4CA37","domain":"pikyoe.eth"}}
```

Both times the same address is used but a different domain is returned => More than one domain can be registered for the same address. In my opinion, returning only one of them is a bug.

File: handlers/ens.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+			} else {
+				returnError = err
+			}
+		} else {
+			data.Domain = domain
+		}
+	} else {
+		returnError = errors.New("not an ens domain or address")
+	}
+	return data, returnError //We always want to return the data if it was a valid address/domain even if there was an error getting data from the node. A valid address might be enough for the caller.
```
[the remark this answers is not in the document]
Comment (MauserBitfly) [reply handlers/ens.go:85]: > I am afraid we will not be able to follow this approach: [#1948 (comment)](https://github.com/gobitfly/eth2-beaconchain-explorer/pull/1948#discussion_r1140096934)
> 
> Furthermore, I have observed the following behavior after starting a local instance of this:
> 
> http://localhost:8081/api/v1/ens/lookup/allice.eth
> 
> ```
> {"status":"OK","data":{"address":"0x20a5bD183b420eBBD1BdA4b8818801b9eFa4CA37","domain":"allice.eth"}}
> ```
> 
> http://localhost:8081/api/v1/ens/lookup/0x20a5bD183b420eBBD1BdA4b8818801b9eFa4CA37
> 
> ```
> {"status":"OK","data":{"address":"0x20a5bD183b420eBBD1BdA4b8818801b9eFa4CA37","domain":"pikyoe.eth"}}
> ```
> 
> Both times the same address is used but a different domain is returned => More than one domain can be registered for the same address. In my opinion, returning only one of them is a bug.

I just tested it now on the ENS page:
You can register multiple ens names with your eth address. And then you can select your 'primary name' and that it exactly the name that we are displaying when you query it with the eth address
image

image


-> So the behavior is correct as we would show the primary ens name for that eth address.

**Comment by MauserBitfly** [comment]
replaced by other PR

**Pull request closed without merging**